adapter/stackdriver/stackdriver.go

```diff
@@ -0,0 +1,295 @@
+// Copyright 2017 the Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stackdriver
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	monitoring "cloud.google.com/go/monitoring/apiv3"
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/googleapis/gax-go"
+	xcontext "golang.org/x/net/context"
+	gapiopts "google.golang.org/api/option"
+	labelpb "google.golang.org/genproto/googleapis/api/label"
+	metricpb "google.golang.org/genproto/googleapis/api/metric"
+	"google.golang.org/genproto/googleapis/api/monitoredres"
+	monitoringpb "google.golang.org/genproto/googleapis/monitoring/v3"
+
+	"istio.io/mixer/adapter/stackdriver/config"
+	"istio.io/mixer/pkg/adapter"
+)
+
+// TODO: implement adapter validation
+// TODO: change batching to be size aware: right now we batch and send data to stackdriver based on only a ticker.
+// Ideally we'd also size our buffer and send data whenever we hit the size limit or config.push_interval time has passed
+// since the last push.
+
+type (
+
+	// Abstracts the creation of the stackdriver client to enable network-less testing.
+	createClientFunc func(*config.Params) (*monitoring.MetricClient, error)
+
+	factory struct {
+		adapter.DefaultBuilder
+		createClient createClientFunc
+	}
+
+	sdinfo struct {
+		ttype string
+		kind  metricpb.MetricDescriptor_MetricKind
+		value metricpb.MetricDescriptor_ValueType
+	}
+
+	// Abstracts over client.CreateTimeSeries for testing
+	pushFunc func(ctx xcontext.Context, req *monitoringpb.CreateTimeSeriesRequest, opts ...gax.CallOption) error
+
+	sd struct {
+		// TODO: remove when env is request scoped
+		env adapter.Env // used for logging
+
+		projectID   string
+		metricInfo  map[string]sdinfo
+		client      *monitoring.MetricClient
+		pushMetrics pushFunc
+		// Stackdriver's SDK doesn't perform batching, though their API supports it.
```

Comment:
side note: we should file an issue for design / impl of universal batching config for generic adapters.